cli/internal/scope/scope.go

```diff
@@ -0,0 +1,238 @@
+package scope
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/mitchellh/cli"
+	"github.com/pkg/errors"
+	"github.com/vercel/turborepo/cli/internal/context"
+	"github.com/vercel/turborepo/cli/internal/fs"
+	"github.com/vercel/turborepo/cli/internal/scm"
+	"github.com/vercel/turborepo/cli/internal/ui"
+	"github.com/vercel/turborepo/cli/internal/util"
+	"github.com/vercel/turborepo/cli/internal/util/filter"
+)
+
+type Opts struct {
+	IncludeDependencies bool
+	IncludeDependents   bool
+	Patterns            []string
+	Since               string
+	Cwd                 string
+	IgnorePatterns      []string
+	GlobalDepPatterns   []string
+}
+
+func ResolvePackages(opts *Opts, scm scm.SCM, ctx *context.Context, tui cli.Ui, logger hclog.Logger) (util.Set, error) {
+	changedFiles, err := getChangedFiles(opts, scm)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note that we do this calculation *before* filtering the changed files.
+	// The user can technically specify both that a file is a global dependency and
+	// that it should be ignored, and currently we treat a change to that file as a
+	// global change.
+	hasRepoGlobalFileChanged, err := repoGlobalFileHasChanged(opts, changedFiles)
+	if err != nil {
+		return nil, err
+	}
+	filteredChangedFiles, err := filterIgnoredFiles(opts, changedFiles)
+	if err != nil {
+		return nil, err
+	}
+
+	changedPackages := make(util.Set)
+	// Be specific with the changed packages only if no repo-wide changes occurred
+	if !hasRepoGlobalFileChanged {
+		changedPackages = getChangedPackages(filteredChangedFiles, ctx.PackageInfos)
+	}
+
+	// Scoped packages
+	// Unwind scope globs
+	scopePkgs, err := getScopedPackages(ctx.PackageNames, opts.Patterns)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid scope")
+	}
+
+	// Filter Packages
+	filteredPkgs := make(util.Set)
+	includeDependencies := opts.IncludeDependencies
+	includeDependents := opts.IncludeDependents
+	// If there has been a global change, changedPackages.Len() will be 0
+	// If both scoped and since are specified, we have to merge two lists:
+	// 1. changed packages that ARE themselves the scoped packages
+	// 2. changed package consumers (package dependents) that are within the scoped subgraph
+	if scopePkgs.Len() > 0 && changedPackages.Len() > 0 {
+		filteredPkgs = scopePkgs.Intersection(changedPackages)
+		for _, changed := range changedPackages {
+			filteredPkgs.Add(changed)
+			includeDependents = true
+
+		}
+		tui.Output(fmt.Sprintf(ui.Dim("• Packages changed since %s in scope: %s"), opts.Since, strings.Join(filteredPkgs.UnsafeListOfStrings(), ", ")))
```

Comment:
in the `--plan` PR we can abstract this into returning some state variable and printing JSON output.